src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1077,28 +1077,31 @@
                            result-binding (conj result-binding)
                            fact-binding (conj fact-binding))
 
-            ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            ; Find children that all parent nodes have.
+            forward-edges (apply s/union (-> (:forward-edges beta-graph)
+                                                (select-keys parent-ids)
+                                                (vals)))
 
             id-to-condition-nodes (:id-to-condition-node beta-graph)
 
-            ;; Find an id for an equivalent node.
-            existing-id (first (for [child-id common-children-ids
-                                     :when (= node (get id-to-condition-nodes child-id))]
-                                 child-id))
-
-            ;; Use the existing id or create a new one.
-            node-id (or existing-id
+            ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
+            ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
+            ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
+            ;; Having different nodes under the same
+            ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
+            update-node->id (fn [m id]
+                              (let [node (get id-to-condition-nodes id)
+                                    prev (get m node)]
+                                (if prev
+                                  (assoc! m node (min prev id))
+                                  (assoc! m node id))))
+
+            node->id (reduce update-node->id
```

Comment:
The map returned here is a transient. Although, `get` does work on that, I think it is something that'd need to be handled with care. I think this is only a 1-time use for now though, so probably ok. I think the cost of going to `persistent!` is fast, but I haven't done a comparison.

Also, as I mentioned above, I'm questioning the need to build this intermediate construct.

Reply:
The [doPersistent](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/jvm/clojure/lang/PersistentHashMap.java#L350) call appears to just build a new persistent map that shares the underlying structure; see the [constructor](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/jvm/clojure/lang/PersistentHashMap.java#L103).  Unless I'm misunderstanding something here the cost of persistent! seems trivial so I'd be inclined to use it if we keep this construct.

Reply:
Yeah `persistent!` has in docs that it is constant time. So it definitely can't be rebuilding a full collection over again. I just wasn't positive what other overhead there was involved. Given the links you posted, looks like very little overhead.